### PR TITLE
Http structured cloudevents

### LINKED ISSUE #38: Validation and support for structured CloudEvents
Need to validate headers in structured CloudEvents calls and automatically toggle between Binary and Structured CloudEvents

### LINKED ISSUE #45: Changelog doesn't reflect v02/v01 removal
## Expected Behavior
Changelog should explain why versions v02 and v01 were removed

## Actual Behavior
Changelog wasn't updated in that PR

## Steps to Reproduce the Problem
View the changelog

## Specifications

- Platform: any

### PR DESCRIPTION
**Pull request by cumason123**
Fixes #38

## Changes
CloudEvent supports structured events
Changelog reflects v02/v01 removals and the new HTTP CloudEvent class

## One line description for the changelog
Built out the CloudEvent class to support both binary and structured events.

- [*] Tests pass
- [*] Appropriate changes to README are included in PR

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,13 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.0.0]
```
Comment (grant) [inline CHANGELOG.md]: Can the Changelog changes be a separate PR?

File: cloudevents/sdk/http_events.py [hunk 2]
```diff
@@ -45,65 +49,110 @@ def __init__(
...
+        # structured data is inside json resp['data']
+        self.data = copy.deepcopy(data) if self.isbinary else \
+            copy.deepcopy(data.get('data', {}))
+
+        self.headers = {
+            **self.required_attribute_values,
+            **self.optional_attribute_values
+        }
+
+    def to_request(
```
Comment (grant) [inline cloudevents/sdk/http_events.py:132]: Can we get a comment for this method?

File: README.md [hunk 3]
```diff
@@ -14,6 +14,57 @@ This SDK current supports the following versions of CloudEvents:
...
+    "id": "structured-event",
+    "source": "README",
+    "data": {"message": "Hello World!"}
+}
+event = CloudEvent(data)
+headers, body = event.to_request()
+
+# POST
+requests.post("<some-url>", json=body, headers=headers)
+```
```
Comment (grant) [inline README.md:64]: Sending CloudEvents should be under receiving CloudEvents in the README as it is relatively less common.

File: README.md [hunk 4]
```diff
@@ -14,6 +14,57 @@ This SDK current supports the following versions of CloudEvents:
 
 Package **cloudevents** provides primitives to work with CloudEvents specification: https://github.com/cloudevents/spec.
 
+Creating minimal CloudEvents with data in version 1.0.0:
```
Comment (grant) [inline README.md:17]: Just say `Sending CloudEvents:`

We can see the version and data in the example and should assume 1.0 anyways.

File: cloudevents/sdk/http_events.py [hunk 5]
```diff
@@ -45,65 +49,110 @@ def __init__(
...
+        data = data if self.isbinary else data_unmarshaller(data)['data']
+        return headers, data
+
+    def __getitem__(self, key):
+        return self.data if key == 'data' else self.headers[key]
+
     @staticmethod
-    def is_binary_cloud_event(headers):
-        for field in base._ce_required_fields:
+    def is_binary_cloud_event(event_handler, headers):
```
Comment (grant) [inline cloudevents/sdk/http_events.py:154]: Is `event_handler` the right word? I don't quite understand it. It seems like it refers to the CloudEvent version.

File: cloudevents/sdk/http_events.py [hunk 6]
```diff
@@ -124,6 +173,10 @@ def detect_event_version(headers, data):
             raise TypeError(f"specversion {specversion} "
                             "currently unsupported")
 
+    @staticmethod
+    def field_name_modifier(field, isbinary):
+        return f"ce-{field}" if isbinary else field
+
```
Comment (grant) [inline cloudevents/sdk/http_events.py:179]: I would inline this in the places it's used.

File: cloudevents/tests/test_http_events.py [hunk 7]
```diff
@@ -43,6 +45,26 @@
     }
 ]
 
+invalid_test_data = [
```
Comment (grant) [inline cloudevents/tests/test_http_events.py:48]: `invalid_cloudevent_request_bodies = [`

File: cloudevents/tests/test_http_events.py [hunk 8]
```diff
@@ -135,12 +197,109 @@ def test_valid_cloud_events(specversion):
...
+@pytest.mark.parametrize("specversion", ['1.0', '0.3'])
+def test_valid_structured_events(specversion):
+    # Test creating multiple cloud events
+    events_queue = []
+    headers = {}
+    num_cloudevents = 30
+    for i in range(num_cloudevents):
+        headers = {
+
+        }
```
Comment (grant) [inline cloudevents/tests/test_http_events.py:288]: Do we need this?

File: samples/http-cloudevents/client.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    # send and print event
+    requests.post(url, headers=headers, json=body)
+    print(
+        f"Sent {event['ce-id']} from {event['ce-source']} with "
+        f"{event['data']}"
+    )
+
+
+if __name__ == "__main__":
+    # expects a url from command line. e.g.
```
Comment (grant) [inline samples/http-cloudevents/client.py:72]: .e.g. ?

Maybe:

`e.g. python3 client.py http://localhost:3000/`

File: README.md (hunk 3 above)
Comment (cumason123) [reply README.md:64]: Unsure of where exactly you want me to move this code sample in the README

Comment (cumason123) [reply README.md]: Do you want me to move this section further down in the README?

Comment (grant) [reply README.md]: I think we should follow-up in a different PR to add a clear way to create a server handler that receives a CE request.
